accepted/0000-workspaces.md

```diff
@@ -0,0 +1,481 @@
+
+The name “workspaces” is already established in the community with both Yarn and Pnpm implementing similar features under that same name so we chose to reuse it for the sake of simplicity to the larger community involved.
+
+## Detailed Explanation
+
+After sourcing feedback from the community, there are 2 major implementations/changes required in the cli in order to provide the feature set that would enable a better management of nested packages.
+- Install: In a workspaces setup users expect to be able to install all nested packages and perform the associated lifecycle scripts from the root-level, it should also be aware of sibling packages that depend on one another and link them appropriately
+- Run some npm subcommands in the context of each nested workspace
+
+## Rationale and Alternatives
+
+First and foremost there’s the alternative of leaving the problem set for userland to solve, there’s already the very popular project Lerna that provides some of these features.
+
+Also available is the alternative of supporting only the install (or bootstrap as Lerna names it) aspect of this proposal, following a less feature-rich approach but one that would still enable the basic goal of improving the user experience of managing multiple child packages but from all the feedback collected during the research phase of this RFC, this alternative is much less desirable to the community of maintainers involved.
+
+## Implementation
+
+### 1. Make npm workspace-aware
+
+We're following the lead of Yarn in supporting the `workspaces` `package.json` field which defines a list of paths, each of these paths may be a workspace itself but it also support globs.
+
+`package.json` example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": {
+        "packages": [
+            "packages/*"
+        ]
+    }
+}
+```
+
+`package.json` shorthand example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": [
+        "packages/*"
+    ]
+}
+```
+
+
+### 2. Installing dependencies across child packages
+
+Change `npm install` ([arborist](https://www.npmjs.com/package/@npmcli/arborist)) behavior to make it install every nested package by default once a valid configuration is defined.
+
+Arborist should also be aware of all nested workspaces and correctly link to an internal workspace should it match the required semver version of an expected dependency anywhere in the installing tree. e.g:
+
+```
+// Given this package.json structure:
+├── package.json { "workspaces": ["dep-a", "dep-b"] }
+├── dep-a
+│   └── package.json { "dependencies": { "dep-b": "^1.0.0" } }
+└── dep-b
+    └── package.json { "version": "1.3.1" }
+
+$ npm install
+
+// Results in this symlinking structure:
+├── node_modules
+│   ├── dep-a -> ./dep-a
+│   └── dep-b -> ./dep-b
```

Comment:
IIUC, this is proposing to link all sub-packages to root-level `node_modules` folder. I find such arrangement as rather problematic, because it allows sub-packages to require other sub-packages that are not listed in the dependencies. Such code will work at runtime when executed inside the monorepo (e.g. when running tests of the package), but will fail once the problematic package is installed from the registry (i.e. when consuming the package in 3rd-party projects).

As an example, consider the example presented in the RFC document, with an additional `dep-c` package with no dependencies and a file `dep-c/index.js` that's calling `require('dep-a')`.

Now when you run `node -e "require('dep-c')"` from monorepo root, all is good. When you run the same command from a project depending on `dep-c`, it will fail with `Error: Cannot find module 'dep-a'`.

In `lerna`, workspace sub-packages are never linked to root-level `node_modules`, the symlinks are always created between the packages only.

```
├── node_modules
├── dep-a
│   └── node_modules
│       └── dep-b -> ../../dep-b
├── dep-b
└── dep-c
```

As a result, `node -e "require('dep-c')"` executed from monorepo root fails.

I understand that hoisting all sub-packages in a shared top-level directory may have benefits. In that case I am proposing to use a different directory name than `node_modules`, e.g. `.workspace_modules`.

```
├── .workspace_modules
│   ├── dep-a -> ./dep-a
│   └── dep-b -> ./dep-b
├── dep-a
│   └── node_modules
│       └── dep-b -> ../../../.workspace_modules/dep-b
└── dep-b
```